Metal capture support for wrapped MTLRenderPipelineState & MTLTexture

Zorro666 opened the pull request:
## Description

Added wrapped `MTLRenderPipelineState` & `MTLTexture`.

Implemented capture serialization for APIs:
`MTLDevice::newRenderPipelineStateWithDescriptor`
`MTLDevice::newTextureWithDescriptor`: two overloads, which use shared helper methods to avoid code duplication. 
In `- (nullable id<MTLTexture>)newTextureWithDescriptor:(MTLTextureDescriptor *)descriptor iosurface:(IOSurfaceRef)iosurface plane:(NSUInteger)plane` get the callstack and do simplistic string comparisons to try to detect calls from `[CAMetalLayer nextDrawable]` and route them to `WrappedMTLDevice::nextDrawableTexture` to use a specific chunk value and string name. It fallbacks to using the `MTLDevice_newTextureWithDescriptor_iosurface` chunk serialisation. Both routes use a shared helper method.

The newTexture APIs use the same serialization code except with different chunk values and string names.

Workaround to prevent Metal capture library assert triggering on the bridge `MTLTexture` object, add protocol `MTLTextureImplementation` to class `ObjCBridgeMTLTexture`

Created C++ structs to mirror the objective-c classes `MTL:: TextureDescriptor` and `MTL:: RenderPipelineDescriptor`. 
These C++ structs are in the `RDMTL` namespace and used for serialization.
There are helper methods to create the C++ structs from ObjC data and vice-versa.

`RDMTL::TextureDescriptor`
`RDMTL::RenderPipelineColorAttachmentDescriptor`
`RDMTL::PipelineBufferDescriptor`
`RDMTL::VertexAttributeDescriptor`
`RDMTL::VertexBufferLayoutDescriptor`
`RDMTL::VertexDescriptor`
`RDMTL::FunctionGroups`
`RDMTL::LinkedFunctions`
`RDMTL::RenderPipelineDescriptor`

Added serialisation for helper types:
`MTL::TextureType`
`MTL::PixelFormat`
`MTL::ResourceOptions`
`MTL::CPUCacheMode`
`MTL::StorageMode`
`MTL::HazardTrackingMode`
`MTL::TextureUsage`
`MTL::TextureSwizzleChannels`
`MTL::TextureSwizzle`
`MTL::BlendFactor`
`MTL::BlendOperation`
`MTL::ColorWriteMask`
`MTL::Mutability`
`MTL::VertexFormat`
`MTL::VertexStepFunction`
`MTL::PrimitiveTopologyClass`
`MTL::TessellationPartitionMode`
`MTL::TessellationFactorFormat`
`MTL::TessellationControlPointIndexType`
`MTL::TessellationFactorStepFunction`
`MTL::Winding`

`MTL::ResourceOptions` is a combined value containing
`MTL::CPUCacheMode`, `MTL::StorageMode`, `MTL::HazardTrackingMode`. This required some custom serialization code: an example of its output:

`MTLResourceCPUCacheModeDefaultCache | MTLResourceStorageModeManaged | MTLResourceHazardTrackingModeDefault`

Includes a bug fix for Metal-CPP where the type was wrong for `MTL::LinkedFunctions::groups`. It was marked as `NS::Array*` but it should be `NS::Dictionary*`.


### Record and Chunks
- When creating `MTLRenderPIpelineState` any `MTLFunction` references in the descriptor have their parent set to be the `MTLRenderPIpelineState` resource record.

Zorro666 committed:
Added wrapped MTLRenderPipelineState & MTLTexture

Implemented capture serialisation for APIs:
MTLDevice::newRenderPipelineStateWithDescriptor
MTLDevice::newTextureWithDescriptor

Workaround to prevent Metal capture library assert triggering on the bridge MTLTexture object, add protocol "MTLTextureImplementation" to class "ObjCBridgeMTLTexture"

Serialisation for helper types:
MTL::TextureDescriptor
MTL::RenderPipelineDescriptor
MTL::RenderPipelineColorAttachmentDescriptor
MTL::PixelFormat
MTL::TextureType
MTL::PrimitiveTopologyClass
MTL::ResourceOptions
MTL::CPUCacheMode
MTL::StorageMode
MTL::HazardTrackingMode
MTL::TextureUsage
MTL::TextureSwizzleChannels
MTL::TextureSwizzle
MTL::ColorWriteMask
MTL::BlendOperation
MTL::BlendFactor
MTL::Winding
MTL::TessellationFactorFormat
MTL::TessellationControlPointIndexType
MTL::TessellationFactorStepFunction
MTL::TessellationPartitionMode

Zorro666 committed:
Fix MTL::LinkedFunctions bug in metal-cpp.h

MTLLinkedFunctions groups is an NS::Dictionary* not an NS::Array*
ie.
@property(readwrite, nonatomic, copy, nullable) NSDictionary<NSString *,NSArray<id<MTLFunction>> *> *groups;

baldurk commented on renderdoc/driver/metal/metal_device.cpp:
```diff
@@ -189,6 +207,122 @@ WrappedMTLLibrary *WrappedMTLDevice::newLibraryWithSource(NS::String *source,
+
+  MTL::RenderPipelineState *realMTLRenderPipelineState;
+  SERIALISE_TIME_CALL(realMTLRenderPipelineState =
+                          Unwrap(this)->newRenderPipelineState(realDescriptor, error));
```
This does not have to be addressed for this PR but something for the future - I presume at some point we will need to check for errors here, and return early without processing if something went wrong?

baldurk commented on renderdoc/driver/metal/metal_device.cpp:
```diff
@@ -320,9 +454,59 @@ bool WrappedMTLDevice::supportsPrimitiveMotionBlur()
+    {
+      CACHE_THREAD_SERIALISER();
+      SCOPED_SERIALISE_CHUNK(chunkType);
+      RDMTL::TextureDescriptor rdDescriptor(descriptor);
```
Duplicate variable declaration of `rdDescriptor` here.

baldurk commented on renderdoc/driver/metal/metal_device_bridge.mm:
```diff
@@ -257,17 +257,32 @@ - (MTLSizeAndAlign)heapBufferSizeAndAlignWithLength:(NSUInteger)length
+  {
+    rdcstr parentCallsite(nsString->utf8String());
+    nextDrawable = (parentCallsite.contains("CAMetalLayer") && parentCallsite.contains("drawable"));
+  }
```
The way this is solved on other platforms is to ensure that the wrapping covers the set of things we need to care about, so we would wrap CAMetalLayer and know the entry point that way. How feasible is that? This feels very brittle, though at least the worst consequence is that the reported call is wrong.

If the whole callstack is examined, will you see the user's call to `nextDrawableTexture`? That seems at least more robust in the event that the internal implementation changes.

baldurk commented on renderdoc/driver/metal/metal_types.cpp:
```diff
@@ -121,4 +123,499 @@ void DoSerialise(SerialiserType &ser, NS::String *&el)
+  SERIALISE_MEMBER(green);
+  SERIALISE_MEMBER(blue);
+  SERIALISE_MEMBER(alpha);
+}
```
By convention `DoSerialise` functions should all be in a `metal_serialise.cpp` file, similar to `metal_stringise.cpp`. At least for public-facing serialisation of API structs and objects.

baldurk commented on renderdoc/driver/metal/metal_types.cpp:
```diff
@@ -121,4 +123,499 @@ void DoSerialise(SerialiserType &ser, NS::String *&el)
+template <typename RDMTL_TYPE, typename MTLARRAY_TYPE, typename MTL_TYPE, int MAX_COUNT>
+static void GetObjcArray(rdcarray<RDMTL_TYPE> &to, MTLARRAY_TYPE *from, bool (*validData)(MTL_TYPE *))
+{
+  MTL_TYPE *objcData[MAX_COUNT];
```
What is the reason for this plain array here? Is there any reason you couldn't `push_back` to the `to` array in the loop below, up to `MAX_COUNT`, optionally with a `reserve` if you wanted? Or if metal's array has a way of querying the length, even better you could reserve/resize using that.

Zorro666 committed:
Refactored WrappedMTLDevice texture helper methods

PR Feedback.
Replaced NewTexture & NewIOSurfaceTextureWithDescriptor with a single combined method Common_NewTexture

Zorro666 committed:
Move DoSerialise* methods to metal_serialise.cpp

PR Feedback.
Moved from metal_types.cpp -> metal_stringise.cpp

Zorro666 committed:
Simplify GetObjcArray implementation

PR Feedback.
Remove optimisation of the array of pointers on the stack
resize the to array when adding a valid object

Zorro666 committed:
Hook CAMetalLayer nextDrawable message

PR Feedback.
Replace the brittle callstack string comparison

Zorro666 committed:
Remove extra rdDescriptor variable

PR Feedback.

Zorro666 committed:
Set nextDrawable TLS data before creating hook

Zorro666 commented on renderdoc/driver/metal/metal_device.h:
```diff
@@ -80,7 +94,12 @@ class WrappedMTLDevice : public WrappedMTLObject
     TypeEnum = eResDevice
   };
 
+  static uint64_t nextDrawableTLSSlot;
```
Was not sure what naming convention to follow here?
could use "s_" to signify static or perhaps "g_" to signify it is a global static

baldurk replied:
I don't tend to have a naming scheme for those, so this is fine as is unless you want to prefix it.

Zorro666 commented on renderdoc/driver/metal/metal_device.h:
```diff
@@ -80,7 +94,12 @@ class WrappedMTLDevice : public WrappedMTLObject
   };
 
+  static uint64_t nextDrawableTLSSlot;
+  static IMP real_CAMetalLayer_nextDrawable;
```
Not sure how many ObjC methods will be swizzled in the end - if there are more then will move them to a helper struct with helper methods and out of WrappedMTLDevice

baldurk commented on renderdoc/driver/metal/metal_device.cpp:
```diff
@@ -42,12 +45,55 @@ WrappedMTLDevice::WrappedMTLDevice(MTL::Device *realMTLDevice, ResourceId objId)
+void WrappedMTLDevice::MTLHookObjcMethods()
+{
+  static bool s_hookObjcMethods = false;
+  if(s_hookObjcMethods)
```
`s_hookObjcMethods` is never set to `true` below here.

baldurk reviewed (changes requested):
One minor thing left but otherwise this looks good to me.

Zorro666 committed:
Adopt g_* naming convention for class static's

WrappedMTLDevice
real_CAMetalLayer_nextDrawable -> g_real_CAMetalLayer_nextDrawable
nextDrawableTLSSlot -> g_nextDrawableTLSSlot

Zorro666 committed:
Set s_hookObjcMethods true in MTLHookObjcMethods

PR Feedback
